File: Chowdren/chowdren/transition.py
from chowdren.common import make_color

FADE_TYPES = {
    'DOOR': 'DOOR',
    'FADE': 'FADE'
}

def write(writer, fade, out):
    fade_dir = 1.0 / (fade.duration / 1000.0)
    if not out:
        fade_dir = -fade_dir
    color = fade.color
    fade_type = 'Transition::%s' % FADE_TYPES.get(fade.name, 'FADE')
    writer.putlnc('manager.set_fade(%s, %s, %s);', fade_type,
                  make_color(color), fade_dir)

File: Chowdren/base/transition.cpp
#include "transition.h"
#include "manager.h"
#include "render.h"

void Transition::draw()
{
    Color c = manager.fade_color;
    float value = manager.fade_value;
    switch (manager.fade_type) {
        case FADE:
            c.set_alpha(int(value * 255));
            Render::draw_quad(0, 0, WINDOW_WIDTH, WINDOW_HEIGHT, c);
            break;
        case DOOR:
            int w = (WINDOW_WIDTH / 2.0f) * value;
            Render::draw_quad(0, 0, w, WINDOW_HEIGHT, c);
            Render::draw_quad(WINDOW_WIDTH - w, 0,
                              WINDOW_WIDTH, WINDOW_HEIGHT, c);
            break;
    }
}